### PR TITLE
add crypto key validation

### PR DESCRIPTION
**Pull request by zreigz**
## Summary
 - Create a `.keyid` file w/ the sha256 digest of the encryption key.
 - Fix the problem with the `workspace.yaml` created outside the git repository
-  Add new command to generate key fingerprint file: `plural crypto key-fingerprint`

## Checklist

- [ ] If required, I have updated the Plural documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a meaningful title and summary to convey the impact of this PR to a user.
- [ ] I have added relevant labels to this PR to help with categorization for release notes.

### CODE REVIEW

File: pkg/provider/aws.go [hunk 1]
```diff
@@ -276,3 +282,7 @@ func GetAwsAccount() (string, error) {
 	}
 	return res.Account, nil
 }
+
+func (aws *AWSProvider) KeyID() string {
```
Comment (michaeljguarino) [inline pkg/provider/aws.go:286]: I don't believe this should be set on the cloud provider actually (I mean we can but it should be orthogonal to it).  Is there a way we can remove the dependency there?

File: cmd/plural/crypto.go [hunk 2]
```diff
@@ -180,13 +182,20 @@ func handleEncrypt(c *cli.Context) error {
...
+	if err != nil {
+		return err
+	}
+	prov, err := provider.GetProvider()
 	if err != nil {
 		return err
 	}
 
-	result, err := crypto.Encrypt(prov, data)
+	if prov.KeyID() != "" && cryptoProv.ID() != prov.KeyID() {
```
Comment (michaeljguarino) [inline cmd/plural/crypto.go:194]: we should probably also do this validation on `plural crypto unlock` too

Comment (michaeljguarino) [reply cmd/plural/crypto.go:194]: If there's a chicken-egg with the `workspace.yaml` being encrypted, then we should maybe create a new file in the repo for the key digest that is outside the encrypted list

Comment (zreigz) [reply cmd/plural/crypto.go:194]: yes, good point

File: pkg/provider/aws.go (hunk 1 above)
Comment (zreigz) [reply pkg/provider/aws.go:286]: we can remove it, especially when we would like to use a new file outside the encrypted list

**Comment by zreigz** [comment]
@michaeljguarino PTAL

File: cmd/plural/init.go [hunk 3]
```diff
@@ -43,9 +43,14 @@ func handleInit(c *cli.Context) error {
 	if err != nil {
 		return err
 	}
-	defer func(prov provider.Provider) {
+	kv, err := crypto.GenerateKeyValidator()
```
Comment (michaeljguarino) [inline cmd/plural/init.go:46]: we should also run this in the setup phase of the cloud shell, and also have an independent crypto command to generate it manually for old repos (i'll maybe call that in the console or something)

Comment (michaeljguarino) [reply cmd/plural/init.go:46]: @zreigz It's in `pkg/server/setup.go`

File: pkg/crypto/key.go [hunk 4]
```diff
@@ -138,3 +138,56 @@ func (k *AESKey) Flush() error {
 
 	return os.WriteFile(getKeyPath(), io, 0644)
 }
+
+func (k *AESKey) ID() string {
+	sha := sha256.Sum256([]byte(k.Key))
```
Comment (michaeljguarino) [inline pkg/crypto/key.go:143]: do we have any utils code for sha'ing currently?  I feel like it's around somewhere

**Comment by michaeljguarino** [comment]
Another thing I realized, and I think you probably covered but we should double-check, you should make sure the sha format here matches what's done in `plural crypto backups list` (you can verify by just creating a backup first w/ `plural crypto backups create`)

**Comment by michaeljguarino** [comment]
So I was able to test it but I realized there's a lot of incompatibilities introduced w/ the new chart, it might be better if we take on the upgrade since it'll be more advanced. It can probably be done in a few days.

**Comment by zreigz** [comment]
> 
Unfortunately, the SHA format is different than for `plural crypto backups` but I am not sure if it's relevant. We compare them in the flight using the same generation function

File: pkg/crypto/key.go (hunk 4 above)
Comment (zreigz) [reply pkg/crypto/key.go:143]: no, we don't have

File: cmd/plural/crypto.go [hunk 5]
```diff
@@ -143,6 +143,11 @@ func (p *Plural) cryptoCommands() []cli.Command {
 			Usage:       "manages backups of your encryption keys",
 			Subcommands: p.backupCommands(),
 		},
+		{
+			Name:   "key-fingerprint",
```
Comment (michaeljguarino) [inline cmd/plural/crypto.go:147]: Perhaps just name this `fingerprint` to be a bit more terse?

File: cmd/plural/init.go [hunk 6]
```diff
@@ -43,20 +43,22 @@ func handleInit(c *cli.Context) error {
...
 		repo, err = scm.Setup()
 		if err != nil {
 			return err
 		}
 		gitCreated = true
-	} else if !git && err != nil {
-		return err
+	}
+	if !git && !gitCreated {
+		return fmt.Errorf("you're not in the git repository")
```
Comment (michaeljguarino) [inline cmd/plural/init.go:55]:
```
you're not in a git repository, either clone one directly or let us set it up for you by rerunning `plural init`
````

**Review by michaeljguarino — APPROVED** [review]
small language nits, but otherwise this is good to go

**Comment by michaeljguarino** [comment]
@zreigz so the difference between the server and here is that we base32 encode it instead of base64 encode it.  If you swap the schemes I think that will keep them aligned

**Comment by zreigz** [comment]
@michaeljguarino I have changed to base32 and they match, thanks

**Pull request merged**